src/PointwiseFunctions/Hydro/EquationsOfState/Tabulated3d.cpp

```diff
@@ -0,0 +1,426 @@
+  p | table_data_;
+}
+
+template <bool IsRelativistic>
+template <class DataType>
+Scalar<DataType>
+Tabulated3D<IsRelativistic>::temperature_from_density_and_energy_impl(
+    const Scalar<DataType>& rest_mass_density,
+    const Scalar<DataType>& specific_internal_energy,
+    const Scalar<DataType>& electron_fraction) const {
+  Scalar<DataType> converted_electron_fraction;
+  Scalar<DataType> log_rest_mass_density;
+
+  Scalar<DataType> log_temperature;
+  Scalar<DataType> temperature;
+
+  temperature = make_with_value<Scalar<DataType>>(rest_mass_density,
+                                                  temperature_lower_bound());
+
+  convert_to_table_quantities(electron_fraction, rest_mass_density, temperature,
+                              converted_electron_fraction,
+                              log_rest_mass_density, log_temperature);
+
+  // Check bounds on eps, note that eps may be negative
+  Scalar<DataType> log_specific_internal_energy = specific_internal_energy;
+
+  if constexpr (std::is_same_v<DataType, double>) {
+    get(log_specific_internal_energy) =
+        std::max(std::min(get(specific_internal_energy),
+                          specific_internal_energy_upper_bound(
+                              get(rest_mass_density), get(electron_fraction))),
+                 specific_internal_energy_lower_bound(get(rest_mass_density),
+                                                      get(electron_fraction)));
+  } else if constexpr (std::is_same_v<DataType, DataVector>) {
+    for (size_t s = 0; s < electron_fraction.size(); ++s) {
+      get(log_specific_internal_energy)[s] = std::max(
+          std::min(get(specific_internal_energy)[s],
+                   specific_internal_energy_upper_bound(
+                       get(rest_mass_density)[s], get(electron_fraction)[s])),
+          specific_internal_energy_lower_bound(get(rest_mass_density)[s],
+                                               get(electron_fraction)[s]));
+    }
+  }
+
+  // Correct for negative eps
+  get(log_specific_internal_energy) -= energy_shift_;
+  get(log_specific_internal_energy) = log(get(log_specific_internal_energy));
+
+  if constexpr (std::is_same_v<DataType, double>) {
+    const auto& log_eps = get(log_specific_internal_energy);
+    const auto& log_rho = get(log_rest_mass_density);
+    const auto& ye = get(converted_electron_fraction);
+
+    // Root-finding appropriate between reference density and maximum density
+    // We can use x=0 and x=x_max as bounds
+    const auto f = [this, log_eps, log_rho, ye](const double log_T) {
+
+      const auto weights = interpolator_.get_weights(log_T, log_rho, ye);
+      const auto interpolated_values =
+          interpolator_.template interpolate<EPS>(weights);
+
+      return log_eps - interpolated_values[0];
+    };
+
+    const auto root_from_lambda = RootFinder::toms748(
+        f, table_log_temperature_.front(),
+        0.99999 * table_log_temperature_.back(), 1.0e-14, 1.0e-15);
```

Comment:
should we have member variables that controls `0.99999` and the tolerances? We can just hard-code the member variable value for now and see if in the future having them as input options would ever be useful, but having fewer magic numbers in the code that aren't assigned to variables makes things easier to understand and maintain. E.g. keeping this line in sync with line 282 below is easy if they are member variables

Reply:
I introduced a new constexpr variable  `upper_bound_tolerance`